logging/impl.go

```diff
@@ -198,6 +245,47 @@ func (imp *impl) shouldLog(logLevel Level) bool {
 }
 
 func (imp *impl) Write(entry *LogEntry) {
+	if imp.registry.DeduplicateLogs.Load() {
+		hashkeyedEntry := entry.HashKey()
+
+		// If we have entered a new recentMessage window, output noisy logs from
+		// the last window.
+		imp.recentMessageMu.Lock()
+		if time.Since(imp.recentMessageWindowStart) > noisyMessageWindowDuration {
+			for stringifiedEntry, count := range imp.recentMessageCounts {
+				if count > noisyMessageCountThreshold {
+					collapsedEntry := imp.recentMessageEntries[stringifiedEntry]
+					collapsedEntry.Message = fmt.Sprintf("Message logged %d times in past %v: %s",
+						count, noisyMessageWindowDuration, collapsedEntry.Message)
+
+					imp.testHelper()
+					for _, appender := range imp.appenders {
+						err := appender.Write(collapsedEntry.Entry, collapsedEntry.Fields)
```

Comment:
The level of the aggregation log that's output here is equivalent to the most recent noisy log's level. You'll see in the test I wrote that this means that if I log 3 identical INFO logs followed by a 4th identical ERROR log, the aggregation log will appear as ERROR level. This is sort of strange, but I'm inclined to not care too much about this case since most identical logs will not differ in level. I suppose this raises the question of whether or not level should be considered as part of identicality but I'll stick to the scope for now.

Reply:
I'm fine with the behavior you described.